src/main/java/gregtech/api/capability/impl/miner/MinerLogic.java

```diff
@@ -0,0 +1,625 @@
+    private static final byte POWER = 5;
+    private static final byte TICK_TOLERANCE = 20;
+    private static final double DIVIDEND = MAX_SPEED * Math.pow(TICK_TOLERANCE, POWER);
+
+    protected final MetaTileEntity metaTileEntity;
+    protected final IMiner miner;
+
+    private final int fortune;
+    private final int speed;
+    private final int maximumRadius;
+
+    private final ICubeRenderer PIPE_TEXTURE;
+
+    private IItemHandlerModifiable outputInventory;
+
+    private final LinkedList<BlockPos> blocksToMine = new LinkedList<>();
+
+    private final AtomicInteger x = new AtomicInteger(Integer.MAX_VALUE);
+    private final AtomicInteger y = new AtomicInteger(Integer.MAX_VALUE);
+    private final AtomicInteger z = new AtomicInteger(Integer.MAX_VALUE);
+    private final AtomicInteger startX = new AtomicInteger(Integer.MAX_VALUE);
+    private final AtomicInteger startZ = new AtomicInteger(Integer.MAX_VALUE);
+    private final AtomicInteger startY = new AtomicInteger(Integer.MAX_VALUE);
+    private final AtomicInteger pipeY = new AtomicInteger(Integer.MAX_VALUE);
+    private final AtomicInteger mineX = new AtomicInteger(Integer.MAX_VALUE);
+    private final AtomicInteger mineZ = new AtomicInteger(Integer.MAX_VALUE);
+    private final AtomicInteger mineY = new AtomicInteger(Integer.MAX_VALUE);
+
+    private int pipeLength = 0;
+    private int currentRadius;
+    private boolean isDone;
+    private boolean isActive = true;
+
+    /**
+     * Creates the general logic for all in-world ore block miners
+     *
+     * @param metaTileEntity the {@link MetaTileEntity} this logic belongs to
+     * @param fortune the fortune amount to apply when mining ores
+     * @param speed the speed in ticks per block mined
+     * @param maximumRadius the maximum radius (square shaped) the miner can mine in
+     */
+    public MinerLogic(MetaTileEntity metaTileEntity, int fortune, int speed, int maximumRadius, ICubeRenderer pipeTexture) {
+        this.metaTileEntity = metaTileEntity;
+        this.miner = (IMiner) metaTileEntity;
+        this.fortune = fortune;
+        this.speed = speed;
+        this.currentRadius = maximumRadius;
+        this.maximumRadius = maximumRadius;
+        this.isDone = false;
+        this.PIPE_TEXTURE = pipeTexture;
+        this.outputInventory = metaTileEntity.getExportItems();
+    }
+
+    /**
+     * Performs the actual mining in world
+     * Call this method every tick in update
+     */
+    public void performMining() {
+        // Needs to be server side
+        if (metaTileEntity.getWorld().isRemote)
+            return;
+
+        // Inactive miners do nothing
+        if (!this.isActive)
+            return;
+
+        // check if mining is possible
+        if (!checkCanMine())
+            return;
+
+        // if the inventory is not full, drain energy etc. from the miner
+        if (!miner.isInventoryFull() && drainStorages(true)) {
```

Comment:
You don't need this `drainStorages(true)` check here, as you already check this condition in the previous `checkCanMine()`